src/librustc_mir/borrow_check/error_reporting.rs

```diff
@@ -521,6 +518,56 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
         err
     }
 
+    fn get_moved_or_unititialized_indexes(
+        &mut self,
+        context: Context,
+        mpi: MovePathIndex,
+    ) -> Vec<MoveOutIndex> {
+        let mir = self.mir;
+
+        let l = context.loc;
+        let mut stack = Vec::new();
+        if context.loc.statement_index == 0 {
+            stack.push(l);
+        } else {
+            stack.push(Location { block: l.block, statement_index: l.statement_index - 1 });
+        }
+
+        let mut visited = FxHashSet();
+        let mut result = vec![];
+
+        'dfs:
+        while let Some(l) = stack.pop() {
+            if !visited.insert(l) {
+                continue;
+            }
+
+            for moi in &self.move_data.loc_map[l] {
+                if self.move_data.moves[*moi].path == mpi {
+                    result.push(*moi);
+                    continue 'dfs;
+                }
+            }
+
+            for ii in &self.move_data.init_loc_map[l] {
+                if self.move_data.inits[*ii].path == mpi {
+                    continue 'dfs;
+                }
```

Comment:
This should probably be instead:

```
use rustc_mir::dataflow::drop_flag_effects;
let mut any_match = false;
drop_flag_effects::for_location_inits(
    self.tcx, 
    self.mir, 
    self.move_data, 
    l,
    |m| if m == mpi { any_match = true; },
);
```

This is a helper function that does the walk for you.

Reply:
That said, there is a subtle case around call returns which I fear this code may get wrong, but I have to think about the example (in particular, `for_local_inits` does not walk the initializations due to a call return).